astropy/cosmology/core.py

```diff
@@ -32,7 +38,265 @@ class CosmologyError(Exception):
+
+    Examples
+    --------
+    The most common use case of ``Parameter`` is to access the corresponding
+    private attribute.
+
+        >>> from astropy.cosmology import LambdaCDM
+        >>> from astropy.cosmology.core import Parameter
+        >>> class Example1(LambdaCDM):
+        ...     param = Parameter(doc="example parameter")
+        ...     def __init__(self, param=15 * u.m):
+        ...         super().__init__(70, 0.3, 0.7)
+        ...         self._param = Parameter.validate(self.__class__.param, param)
+        >>> Example1.param
+        <Parameter 'Example1.param' at ...
+        >>> Example1.param.default
+        <Quantity 15. m>
+        >>> Example1.param.unit
+        Unit("m")
+
+        >>> ex = Example1(param=12357)
+        >>> ex.param
+        <Quantity 12357. m>
+
+    ``Parameter`` also supports custom ``getter`` methods.
+    :attr:`~astropy.cosmology.FLRW.m_nu` is a good example.
+
+        >>> import astropy.units as u
+        >>> class Example2(LambdaCDM):
+        ...     param = Parameter(doc="example parameter", unit="m")
+        ...     def __init__(self, param=15):
+        ...         super().__init__(70, 0.3, 0.7)
+        ...         self._param = Parameter.validate(self.__class__.param, param)
+        ...     @param.getter
+        ...     def param(self):
+        ...         return self._param << u.km
+
+        >>> ex2 = Example2(param=12357)
+        >>> ex2.param
+        <Quantity 12.357 km>
+
+    .. doctest::
+       :hide:
+
+       >>> from astropy.cosmology.core import _COSMOLOGY_CLASSES
+       >>> _ = _COSMOLOGY_CLASSES.pop(Example1.__qualname__)
+       >>> _ = _COSMOLOGY_CLASSES.pop(Example2.__qualname__)
+    """
+
+    def __init__(self, fget=None, doc=None, *, unit=None, equivalencies=[], fmt=".3g"):
+        # modeled after https://docs.python.org/3/howto/descriptor.html#properties
+        self.fget = fget
+        self.__doc__ = fget.__doc__ if (doc is None and fget is not None) else doc
+        self._fmt = str(fmt)
+        # units stuff
+        self._unit = u.Unit(unit) if unit is not None else None
+        self._equivalencies = equivalencies
+
+        # compatibility with other descriptors:
+        self.__name__ = getattr(fget, "__name__", None)
+        self._is_fget_descriptor = hasattr(fget, "fget")
+        # TODO! better detection. Problem is methods have a `__get__`.
+
+    def __set_name__(self, objcls, name):
+        # attribute name
+        self._attr_name = name
+        self._attr_name_private = "_" + name
+
+        # update __name__, if not already set
+        self.__name__ = self.__name__ or name
+
+        # default value (done here b/c 1st access to 'objcls')
+        nullparam = inspect.Parameter('None', 1, default=inspect._empty)
+        param = objcls._init_signature.parameters.get(name, nullparam)
+        self._default = param.default if param.default is not inspect._empty else None
```

Comment:
> Wouldn't it make more sense to just get it in the property below?

It would. I just put it there and it looks much nicer.

Reply:
> I'm actually a bit confused whether it is needed to know about the default at all!?

Strictly speaking, not yet. But 1) in order to have the nice ``__repr__`` and auto-units I needed the class, and having the default value came free and 2) ``to_format`` also works when called from the class and I was thinking ``Cosmology.to_format("astropy.model", "<method name>")`` would return an uninitialized ``astropy.modeling.Model`` with the correct defaults set. When called from a Cosmology instance it would return a Model instance.

Reply:
But I can hold off on adding ``default`` until that PR. This PR already does enough.

Reply:
See https://github.com/astropy/astropy/pull/12136#issuecomment-915618900. I think this problem is tricky and can be tackled at a later date.

Reply:
Resolving because moving to a future PR.